From: Li Yu <bingtian.ly@taobao.com>
Subject: rt-stat: skip the TCP connection that SYN-ACK is ever retransmitted.
Patch-mainline: In house
References: 

Skip these connections since they prolong measured latency wrongly.

Acked-by: 
Signed-off-by: Li Yu <bingtian.ly@taobao.com>

--- a/net/ipv4/rt_stat.c	2012-06-19 12:29:45.053273773 -0400
+++ b/net/ipv4/rt_stat.c	2012-06-19 12:30:14.245273028 -0400
@@ -376,7 +376,9 @@
 
 	preempt_disable();
 	cpu = smp_processor_id();
-	if (req->ts_incoming && AF_INET == req->rsk_ops->family) {
+	if (req->ts_incoming
+			&& !req->retrans
+			&& AF_INET == req->rsk_ops->family) {
 		tcp_rt_stat(cpu, req);
 		req->ts_incoming = 0;
 	}
