From: Fengguang Wu <fengguang.wu@intel.com>
Subject: writeback: delay periodic work on wrap
Date: Fri Dec 16 19:19:16 CST 2011
Patch-mainline: not-upstream-yet

This guarantees some break time on busy overwriters.

Signed-off-by: Wu Fengguang <fengguang.wu@intel.com>
Signed-off-by: Robin Dong <sanbai@taobao.com>
---
Index: linux-2.6.32-220.2.1.el5/mm/page-writeback.c
===================================================================
--- linux-2.6.32-220.2.1.el5.orig/mm/page-writeback.c	2012-02-24 11:53:01.168748022 +0800
+++ linux-2.6.32-220.2.1.el5/mm/page-writeback.c	2012-02-24 11:53:08.376783662 +0800
@@ -905,6 +905,7 @@
 		nr_pages = pagevec_lookup_tag(&pvec, mapping, &index, tag,
 			      min(end - index, (pgoff_t)PAGEVEC_SIZE-1) + 1);
 		if (nr_pages == 0) {
+			mapping->host->dirtied_when = jiffies;
 			done_index = 0;
 			break;
 		}
